server.ts

```diff
@@ -201,62 +212,191 @@ export function app() {
- * The cache control value can be configured in the environments file and defaults to max-age=60
+ * Adds a Cache-Control HTTP header to the response.
+ * The cache control value can be configured in the config.*.yml file
+ * Defaults to max-age=604,800 seconds (1 week)
  */
-function cacheControl(req, res, next) {
+function addCacheControl(req, res, next) {
   // instruct browser to revalidate
-  res.header('Cache-Control', environment.cache.control || 'max-age=60');
+  res.header('Cache-Control', environment.cache.control || 'max-age=604800');
   next();
 }
 
+/*
+ * Initialize server-side caching of pages rendered via SSR.
+ */
+function initCache() {
+  if (cacheEnabled()) {
+    // Initialize a new "least-recently-used" item cache (where least recently used items are removed first)
+    // See https://www.npmjs.com/package/lru-cache
+    cache = new LRU( {
+      max: environment.cache.serverSide.max || 100,            // 100 items in cache maximum
+      ttl: environment.cache.serverSide.timeToLive || 15 * 60 * 1000, // 15 minute cache
+      allowStale: true // If object is found to be stale, return stale value before deleting
+    });
+  }
+}
+
+/**
+ * Return whether server side caching is enabled in configuration.
+ */
+function cacheEnabled(): boolean {
+  // Caching is only enabled is SSR is enabled AND
+  // "serverSide.max" setting is greater than zero
+  return environment.universal.preboot && environment.cache.serverSide.max && (environment.cache.serverSide.max > 0);
+}
+
+/**
+ * Check if the currently requested page is in our server-side, in-memory cache.
+ * Caching is ONLY done for SSR requests. Pages are cached base on their path (e.g. /home or /search?query=test)
+ */
+function cacheCheck(req, res, next) {
+  let cacheHit = false;
+  let debug = false; // Enable to see cache hits & re-rendering logs
+
+  // Only check cache if cache enabled & NOT authenticated.
+  // NOTE: Authenticated users cannot use the SSR cache. Cached pages only show data available to anonymous users.
+  // Only public pages can currently be cached, as the cached data is not user-specific.
+  if (cacheEnabled() && !isUserAuthenticated(req)) {
+    const key = getCacheKey(req);
+
+    // Check if this page is in our cache
+    let cachedCopy = cache.get(key);
+    if (cachedCopy) {
+      cacheHit = true;
+      res.locals.ssr = true;  // mark response as SSR (enables text compression)
+      if (debug) { console.log(`CACHE HIT FOR ${key}`); }
+      // return page from cache to user
+      res.send(cachedCopy);
+
+      // Check if cached copy is expired (in this sitution key will now be gone from cache)
+      if (!cache.has(key)) {
+        if (debug) { console.log(`CACHE EXPIRED FOR ${key} Re-rendering...`); }
+        // Update cached copy by rerendering server-side
+        // NOTE: Cached copy was already returned to user above. So, this re-render is just to prepare for next user.
+        serverSideRender(req, res, false);
+      }
```

Comment:
Wouldn't this mean we can return cached copies that are older than 15 minutes? Perhaps it's best to rerender before returning the response

Reply:
@artlowel : Yes, it's possible that the cached copy could be older than 15 minutes... it's just guaranteed to be at least 15 mins old (based on what you set `timeToLive` to).

The intention of this behavior is to ensure the SSR is occurring _behind the scenes_ as much as possible.  So, that users rarely need to wait on the SSR processing.  In this example, the user who encountered the "stale" page does still see the cached copy, instead of that user having to wait for SSR to generate a new copy.  

Either approach can work here (and there are pros/cons to both approaches). But my goal was to see if we could find a way that users *rarely* had to wait on SSR.